src/crimson/os/seastore/transaction_manager.h

```diff
@@ -424,6 +424,140 @@ class TransactionManager : public ExtentCallbackInterface {
     });
   }
 
+  /**
+   * overwrite_pin
+   *
+   * Overwrite extent inside the original extent.
+   * Return the tuple of splited extents and overwrite extent.
+   * One of left pin or right pin can be nullptr.
+   */
+  using overwrite_pin_iertr = base_iertr;
+  template <typename T>
+  using overwrite_pin_ret = overwrite_pin_iertr::future<
+    std::tuple<LBAMappingRef, TCachedExtentRef<T>, LBAMappingRef>>;
+  template <typename T>
+  overwrite_pin_ret<T> overwrite_pin(
+    Transaction &t,
+    LBAMappingRef &&pin,
+    extent_len_t overwrite_offset,
+    extent_len_t overwrite_len,
+    ceph::bufferlist bl) {
+    LOG_PREFIX(TransactionManager::overwrite_pin);
+    // FIXME: paddr can be absolute and pending
+    ceph_assert(pin->get_val().is_absolute());
+
+    auto original_laddr = pin->get_key();
+    auto original_paddr = pin->get_val();
+    auto original_length = pin->get_length();
+    auto l_laddr = original_laddr;
+    auto l_paddr = original_paddr;
+    auto l_len = overwrite_offset;
+    auto o_laddr = original_laddr + overwrite_offset;
+    auto o_paddr = original_paddr.add_offset(overwrite_offset);
+    auto o_len = overwrite_len;
+    auto r_laddr = original_laddr + l_len + o_len;
+    auto r_paddr = original_paddr.add_offset(l_len + o_len);
+    auto r_len = original_length - l_len - o_len;
+    SUBDEBUGT(seastore_tm, "original laddr: {}, paddr: {}, length: {}"
+      " left new laddr: {}, paddr: {}, length: {}"
+      " overwrite new laddr: {}, paddr: {}, length: {}"
+      " right new laddr: {}, paddr: {}, length: {}",
+      t, original_laddr, original_paddr, original_length,
+      l_laddr, l_paddr, l_len,
+      o_laddr, o_paddr, o_len,
+      r_laddr, r_paddr, r_len);
+    ceph_assert(original_length > o_len);
+    ceph_assert(o_laddr >= original_laddr);
```

Comment:
`o_laddr >= original_laddr` looks meaningless given `overwrite_offset` is unsigned.